execution/evm/execution.go

```diff
@@ -355,80 +383,96 @@ func (c *EngineClient) ExecuteTxs(ctx context.Context, txs [][]byte, blockHeight
 		}
 
-		payloadID = forkchoiceResult.PayloadID
+		newPayloadID = forkchoiceResult.PayloadID
 		return nil
 	}, MaxPayloadStatusRetries, InitialRetryBackoff, "ExecuteTxs forkchoice")
 	if err != nil {
 		return nil, 0, err
 	}
 
-	// get payload
-	var payloadResult engine.ExecutionPayloadEnvelope
-	err = c.engineClient.CallContext(ctx, &payloadResult, "engine_getPayloadV4", *payloadID)
-	if err != nil {
-		return nil, 0, fmt.Errorf("get payload failed: %w", err)
+	// Save ExecMeta with payloadID for crash recovery (Stage="started")
+	// This allows resuming the payload build if we crash before completing
+	if c.store != nil {
+		c.saveExecMeta(ctx, blockHeight, timestamp.Unix(), newPayloadID[:], nil, nil, txs, store.ExecStageStarted)
 	}
 
-	// Submit payload with retry logic for SYNCING status
-	var newPayloadResult engine.PayloadStatusV1
-	err = retryWithBackoffOnPayloadStatus(ctx, func() error {
-		err := c.engineClient.CallContext(ctx, &newPayloadResult, "engine_newPayloadV4",
-			payloadResult.ExecutionPayload,
-			[]string{},          // No blob hashes
-			common.Hash{}.Hex(), // Use zero hash for parentBeaconBlockRoot (same as in payload attributes)
-			[][]byte{},          // No execution requests
-		)
-		if err != nil {
-			return fmt.Errorf("new payload submission failed: %w", err)
-		}
-
-		// Validate payload status
-		if err := validatePayloadStatus(newPayloadResult); err != nil {
-			c.logger.Warn().
-				Str("status", newPayloadResult.Status).
-				Str("latestValidHash", newPayloadResult.LatestValidHash.Hex()).
-				Interface("validationError", newPayloadResult.ValidationError).
-				Uint64("blockHeight", blockHeight).
-				Msg("engine_newPayloadV4 returned non-VALID status")
-			return err
-		}
-		return nil
-	}, MaxPayloadStatusRetries, InitialRetryBackoff, "ExecuteTxs newPayload")
-	if err != nil {
-		return nil, 0, err
-	}
+	// 4. Process the payload (get, submit, finalize)
+	return c.processPayload(ctx, *newPayloadID, txs)
+}
 
-	// forkchoice update
-	blockHash := payloadResult.ExecutionPayload.BlockHash
-	err = c.setFinal(ctx, blockHash, false)
-	if err != nil {
-		return nil, 0, err
+// setHead updates the head block hash without changing safe or finalized.
+// This is used when reusing an existing block (idempotency check).
+func (c *EngineClient) setHead(ctx context.Context, blockHash common.Hash) error {
+	c.mu.Lock()
+	c.currentHeadBlockHash = blockHash
+	// Note: safe and finalized are NOT updated - they advance separately via derivation
+	args := engine.ForkchoiceStateV1{
+		HeadBlockHash:      c.currentHeadBlockHash,
+		SafeBlockHash:      c.currentSafeBlockHash,
+		FinalizedBlockHash: c.currentFinalizedBlockHash,
 	}
+	c.mu.Unlock()
 
-	return payloadResult.ExecutionPayload.StateRoot.Bytes(), payloadResult.ExecutionPayload.GasUsed, nil
+	return c.doForkchoiceUpdate(ctx, args, "setHead")
 }
 
 func (c *EngineClient) setFinal(ctx context.Context, blockHash common.Hash, isFinal bool) error {
```

Comment:
bit weird to have two private setFinal and setFinalWithHeight. Can we merge them?